Add function to generate the configuration file

saimn opened the pull request:
Ref  #10146. This was fun and quick enough to do, but I had to hard-code the list of modules to import, not sure if there is a more clever (and simple) solution ?

Result: https://gist.github.com/saimn/b01966dbc6cef2c8c849ac4b933e84d3
This is almost the same as `astropy.cfg`, the order of subpackages is slightly different (alphabetical with the function here), and the only thing missing is the manual subsections like ` ### CORE DATA STRUCTURES AND TRANSFORMATIONS`.

Next steps:
- Find a way to include the output in the docs
- Should this be used to update `astropy.cfg` occasionally, or replace it ?
- Replace `astropy.cfg` is not so easy since it is used as a template for the default config file, which is checking while importing astropy. But the solution here takes a bit of time since it needs to import all modules with config items...

EDIT: Fix #10146

bsipocz commented on astropy/config/configuration.py:
```diff
@@ -570,6 +574,57 @@ def get_config(packageormod=None, reload=False, rootname=None):
+
+    # First, we need to import all modules containing a ConfigNamespace
+    # subclass. So this list must be kept up-to-date.
+    modules_to_import = [
```
could we import everything and check whether there are any `ConfigNamespace` classes used? 

I'm otherwise afraid that this won't be updated when a new usage is introduced

pllim replied:
There is no standard place to store `conf` in. The only way is either hardcode or scrape the code.

mhvk replied:
I'm with @bsipocz here. Not sure it is that hard: we'd only have to recursively import modules - e.g., https://stackoverflow.com/questions/3365740/how-to-import-all-submodules
Alternatively, if we somehow put this at the end of a full astropy test run, we can be sure it is complete (anything that is not tested should not be included!)

bsipocz replied:
I already have the importing everything hack for a test, I suppose that can be easily reused. I'll dig it up in a bit

pllim commented:
A quick grep shows you where they are defined:
```
astropy/config/configuration.py:92:        conf = Conf()
astropy/io/fits/__init__.py:62:conf = Conf()
astropy/io/votable/__init__.py:36:conf = Conf()
astropy/logger.py:87:conf = Conf()
astropy/nddata/__init__.py:51:conf = Conf()
astropy/samp/__init__.py:39:conf = Conf()
astropy/table/jsviewer.py:30:conf = Conf()
astropy/table/__init__.py:45:conf = Conf()  # noqa
astropy/units/quantity.py:59:conf = Conf()
astropy/utils/data.py:95:conf = Conf()
astropy/utils/iers/iers.py:145:conf = Conf()
astropy/visualization/wcsaxes/__init__.py:41:conf = Conf()
astropy/__init__.py:134:conf = Conf()
```

Re: Replacing `astropy.cfg` -- Let's not get too ambitious right now, especially if we decide on not using it by default anymore as mentioned in https://github.com/astropy/astropy/pull/10090#issuecomment-608518871 . Also even if we want to replace it, how is it going to be generated in a way that is compatible with the various way that people might be installing it? Easiest option is maybe add a step for the release maintainer to do, as with updating the leap seconds table ((ref)[https://docs.astropy.org/en/latest/development/releasing.html]).

pllim commented on astropy/config/configuration.py:
```diff
@@ -570,6 +574,57 @@ def get_config(packageormod=None, reload=False, rootname=None):
         return cobj
 
 
+def generate_config(filename=None):
```
Other packages use the configuration system too, so should be more explicit in the name that this won't work for them. Or find a way to let user customize the list of modules.
```suggestion
def generate_astropy_config(filename=None):
```

pllim commented on astropy/config/configuration.py:
```diff
@@ -570,6 +574,57 @@ def get_config(packageormod=None, reload=False, rootname=None):
+    wrapper = TextWrapper(initial_indent="## ", subsequent_indent='## ',
+                          width=78)
+
+    fp = open(filename, 'w') if filename else sys.stdout
```
Just a personal opinion: I find `sys.stdout` not useful except for debugging. A more useful feature is to return the stuff as one string (with line breaks), a list of strings (without line breaks), or dictionary. Then user can do whatever user wants with that output (write to file, print to screen, etc). Also easier to test.

saimn replied:
The idea is that stdout may be useful for docs inclusion, but we will see. Otherwise I think it would be useful to save this config file in its default location, and print where this is.

pllim replied:
I would prefer using `.. literalinclude::` over `stdout`.

mhvk replied:
I think @pllim is right - better to just have the ability to get a list of strings, or an `StringIO()` object, even for creating a documentation page.

saimn replied:
Ok, I changed to accept file objects instead.

mhvk commented:
This looks very good! My own sense would be to try to aim to have it auto-included in the documentation. If that's done at the end, we don't even have to worry about importing astropy modules...

saimn commented:
Hmm, importing everything has something interesting side effects, like showing a pandas :grinning: (and showing deprecation warning, and crashing on `convolve._convolve`):
```
Importing astropy.table.meta
Importing astropy.table.np_utils
Importing astropy.table.operations
Importing astropy.table.pandas

░░░░░░░░▄██▄░░░░░░▄▄░░
░░░░░░░▐███▀░░░░░▄███▌
░░▄▀░░▄█▀▀░░░░░░░░▀██░
░█░░░██░░░░░░░░░░░░░░░
█▌░░▐██░░▄██▌░░▄▄▄░░░▄
██░░▐██▄░▀█▀░░░▀██░░▐▌
██▄░▐███▄▄░░▄▄▄░▀▀░▄██
▐███▄██████▄░▀░▄█████▌
▐████████████▀▀██████░
░▐████▀██████░░█████░░
░░░▀▀▀░░█████▌░████▀░░
░░░░░░░░░▀▀███░▀▀▀░░░░
Importing astropy.table.pprint
Importing astropy.table.row
Importing astropy.table.scripts
Importing astropy.table.scripts.showtable
Importing astropy.table.serialize
Importing astropy.table.setup_package
Importing astropy.table.soco
Importing astropy.table.sorted_array
Importing astropy.table.table
Importing astropy.table.table_helpers
Importing astropy.table.tests
Importing astropy.tests
Importing astropy.tests.plugins.display
WARNING: AstropyDeprecationWarning: The astropy.tests.plugins.display plugin has been deprecated. See the pytest-astropy-header documentation for information on migrating to using pytest-astropy-header to customize the pytest header. [astropy.tests.plugins.display]
Importing astropy.time
Importing astropy.time.core
Importing astropy.time.formats
Importing astropy.time.tests
Importing astropy.time.utils
...
```

saimn commented:
Interesting, maybe some easter eggs, or my system is infected by pandas ... :panda_face: 
```
In [1]: import pandas                                                                          

In [2]: import astropy.table.pandas                                                            

░░░░░░░░▄██▄░░░░░░▄▄░░
░░░░░░░▐███▀░░░░░▄███▌
░░▄▀░░▄█▀▀░░░░░░░░▀██░
░█░░░██░░░░░░░░░░░░░░░
█▌░░▐██░░▄██▌░░▄▄▄░░░▄
██░░▐██▄░▀█▀░░░▀██░░▐▌
██▄░▐███▄▄░░▄▄▄░▀▀░▄██
▐███▄██████▄░▀░▄█████▌
▐████████████▀▀██████░
░▐████▀██████░░█████░░
░░░▀▀▀░░█████▌░████▀░░
░░░░░░░░░▀▀███░▀▀▀░░░░
```

bsipocz commented:
I have this stupid little script here: https://github.com/astropy/ci-helpers/blob/master/utils/import_submodules.py that maybe of use here, too.

(I use it in a WIP PR: https://github.com/astropy/astropy/pull/7952/files#diff-b91f3d5bd63fcd17221b267e851608e8R94)

saimn commented:
@bsipocz - thanks, I just pushed something using `pkgutil.walk_packages`, and just skipping test packages.

saimn commented:
So I added a very simple Sphinx extension to include directly the generated config in the docs:
https://64628-2081289-gh.circle-artifacts.com/0/docs/_build/html/config/astropy_config.html#astropy-config-file
with a link in https://64628-2081289-gh.circle-artifacts.com/0/docs/_build/html/config/index.html#getting-started

Should this extension go in sphinx-astropy ?
Should I update the `astropy.cfg` file for now ?

pllim commented:
> Should this extension go in sphinx-astropy ?

This extension only works for the core library, right? Should it just stay here like it is already in the PR?

saimn commented:
Fine by me :+1:

bsipocz commented:
I see that we might want to use something like this for astroquery where we have configs for all subpackages, but keeping it in core for now sounds reasonable.

pllim commented on astropy/config/configuration.py:
```diff
@@ -570,6 +575,64 @@ def get_config(packageormod=None, reload=False, rootname=None):
+        If None, the default configuration path is taken from `get_config`.
+
+    """
+    import astropy
```
I guess it wouldn't be a lot more work to generalize this part for any package (like `astroquery`) to use? If generalize, then I think it does make sense to live in `sphinx_astropy`. 🤔

saimn replied:
Yes, it should be pretty easy to generalize, I can have a look later.

mhvk commented on astropy/config/configuration.py:
```diff
@@ -570,6 +575,64 @@ def get_config(packageormod=None, reload=False, rootname=None):
+    if filename is None:
+        filename = get_config('astropy').filename
+
+    if isinstance(filename, str):
```
One neat way to do this, allowing you to get rid of the `try/finally` is:
```
import contextlib

with contextlib.ExitStack() as stack:
    if isinstance(filename, str):
        fp = stack.enter_context(open(filename, 'w'))
```

mhvk commented on docs/ext/astropy_config.py:
```diff
@@ -0,0 +1,27 @@
+        generate_astropy_config(buf)
+        text = buf.getvalue()
+        node = nodes.literal_block(text, text)
+        # self.add_name(node)
```
Left-over from a trial?

saimn replied:
Yep, I wanted to see what it does but forgot about it ;)

saimn commented:
I addressed the comments, making the function generic, and I will move the sphinx directive to sphinx-astropy, but he windows build fails:
```
Windows fatal exception: access violation
```
https://travis-ci.org/github/astropy/astropy/jobs/678447655

Adding debug flag to pytest, this seems to be because of the pandas ...
```
=================================== ERRORS ====================================
_ ERROR collecting .tox/py37-test-alldeps/lib/site-packages/astropy/table/pandas.py _
..\..\.tox\py37-test-alldeps\lib\site-packages\astropy\table\pandas.py:29: in <module>
    display.display(dhtml)
..\..\.tox\py37-test-alldeps\lib\site-packages\IPython\core\display.py:282: in display
    print(*objs)
..\..\.tox\py37-test-alldeps\lib\site-packages\colorama\ansitowin32.py:41: in write
    self.__convertor.write(text)
..\..\.tox\py37-test-alldeps\lib\site-packages\colorama\ansitowin32.py:162: in write
    self.write_and_convert(text)
..\..\.tox\py37-test-alldeps\lib\site-packages\colorama\ansitowin32.py:190: in write_and_convert
    self.write_plain_text(text, cursor, len(text))
..\..\.tox\py37-test-alldeps\lib\site-packages\colorama\ansitowin32.py:195: in write_plain_text
    self.wrapped.write(text[start:end])
c:\tools\miniconda3\envs\test\lib\encodings\cp1252.py:19: in encode
    return codecs.charmap_encode(input,self.errors,encoding_table)[0]
E   UnicodeEncodeError: 'charmap' codec can't encode characters in position 2-23: character maps to <undefined>
!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
```
https://travis-ci.org/github/astropy/astropy/jobs/678645380

pllim commented on astropy/config/configuration.py:
```diff
@@ -570,6 +577,68 @@ def get_config(packageormod=None, reload=False, rootname=None):
+        for mod in pkgutil.walk_packages(path=package.__path__,
+                                         prefix=package.__name__ + '.'):
+
+            if mod.module_finder.path.endswith('tests'):
```
Just in case?
```suggestion
            if mod.module_finder.path.endswith(('tests', 'test')):
```

saimn replied:
Done.

pllim commented on astropy/config/configuration.py:
```diff
@@ -570,6 +577,68 @@ def get_config(packageormod=None, reload=False, rootname=None):
+                    # module name, but not if this is the root package...
+                    if item.module != pkgname:
+                        modname = item.module.replace(f'{pkgname}.', '')
+                        fp.write(f"[{modname}]\n\n")
```
Is it safer to use `os.linesep` instead of `"\n"`?

saimn replied:
`open` takes care of this by default:
> * On input, if newline is None, universal newlines mode is
  enabled. Lines in the input can end in '\n', '\r', or '\r\n', and
  these are translated into '\n' before being returned to the
  caller. [...]
> * On output, if newline is None, any '\n' characters written are
>  translated to the system default line separator, os.linesep.

mhvk replied:
Ah, that's so good to know!

pllim commented on astropy/config/tests/test_configs.py:
```diff
@@ -129,6 +130,25 @@ def test_config_file():
+    with open(outfile) as fp:
+        conf2 = fp.read()
+
+    for c in (conf, conf2):
```
Maybe a comment to explain why these specific lines are selected for testing? Were they arbitrarily selected or were there reasons?

saimn replied:
Just a random selection :smile:

saimn replied:
Added a comment.

saimn commented on astropy/astropy.cfg:
```diff
@@ -1,144 +1,169 @@
-# -*- coding: utf-8 -*-
```
It was mostly an issue with Python 2, but with Python 3 it should be ok:
> If no encoding declaration is found, the default encoding is UTF-8.
https://docs.python.org/3/reference/lexical_analysis.html#encoding-declarations

bsipocz commented:
Yes, merging this should wait until the sphinx-astropy release is out. Alternatively we can add the dev version as dependency in the meantime and wait with the release until the end of the RC testing period. Either would work equally well for me.

There is one more thing, we'll need to add a new minimum sphinx-astropy version requirement, otherwise the docs build will fail with the missing directive error.

astrofrog commented:
I've just released sphinx-astropy 1.3! Can you try adding that as a minimum requirement if we are going to be relying on this new plugin? (and this will also trigger a CI build)

saimn commented:
It works: https://66219-2081289-gh.circle-artifacts.com/0/docs/_build/html/config/astropy_config.html#astropy-config-file

bsipocz commented:
Thank you @saimn, this looks awesome now!
